Amazon.IonDotnet/Internals/Binary/ManagedBinaryWriter.cs

```diff
@@ -162,75 +161,91 @@ public override void SetFieldName(string name)
 
         public override void SetFieldNameSymbol(SymbolToken symbol)
         {
+            this.ThrowIfDisposed();
             var token = this.InternSymbol(symbol);
             this.userWriter.SetFieldNameSymbol(token);
         }
 
         public override void StepIn(IonType type)
         {
+            this.ThrowIfDisposed();
+
             // TODO implement top-level symbol table
             this.userWriter.StepIn(type);
         }
 
         public override void StepOut()
         {
+            this.ThrowIfDisposed();
```

Comment:
In the methods where the managed writer delegates directly to the raw writer, let's drop the `ThrowIfDisposed` check, because it's redundant.

Reply:
I disagree here.

It's good practice to throw the exception in the class itself when it's disposed, instead of delegating it to child classes that are disposed upon disposal of the parent class. 

It can be confusing because then the exception message will say it comes from the managed writer being disposed, which is more of a hidden implementation detail. 

In this case, we currently can handle understanding the entire flow to say that we can delegate it to the raw writer, but the code paths can become complex, and in some cases the delegated exception is hit, and some cases it's not. It's better to ensure consistency.

Reply:
Both `ManagedBinaryWriter` and `RawBinaryWriter` are internal implementation details. The user only interacts with the `IIonWriter` interface. The user should not depend on a particular stack trace being included with an `ObjectDisposedException`. We should probably modify the message for both `ManagedBinaryWriter` and `RawBinaryWriter` to be something like "Dispose() called on an IIonWriter that was previously disposed."

I agree that some of the methods in `ManagedBinaryWriter` are more complex than a simple delegation to `RawBinaryWriter`. In those cases, it's fine to keep the check within `ManagedBinaryWriter`. But in the cases where we know that the method in `ManagedBinaryWriter` is only a direct delegation to `RawBinaryWriter`, we should not require a redundant double-check in those cases. If we're worried that in the future we may accidentally remove those checks from `RawBinaryWriter`, we can add unit tests that make sure the exception is thrown from each method after disposal.

Reply:
I've changed the error string message and allowed delegation to the RawBinaryWriter for all functions except one.

The commits are also squashed now, since (hopefully) this was the last comment to address.